### PR TITLE
Simplified Stack Creation

### PR DESCRIPTION
**Pull request by rnzsgh**
This PR adds wrapper stacks for one-click installation of the various components. The CFN templates are being hosted in the startup account in the awslabs-startup-kit-templates-deploy-v1 bucket. The deploy.sh script in bin was added to assist in updating the templates.

The foundation for fully automated regional testing was added in bin/test.py. The bastion host AMIs were also updated to the latest versions, the EB application stacks were updated to the latest versions and support for us-west-3 was added.

The base/core CFN templates were moved to the "templates" directory, so we will need to update blog posts that directly reference these files will need to be updated in tandem.

Manually tested the Fargate stacks and did manual testing of the EB, VPC and bastion stacks in various regions and numerous automated tests in all of the supported regions.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -144,3 +144,122 @@ assigned to the "AppSecurityGroup" so they can talk to the database as
 well as the load balancer, and receive SSH traffic from the bastion host.
 
 
+### Deploy
+
+New services are not always available in all AWS Regions when they are released. Please consult the
+[Region Table](https://aws.amazon.com/about-aws/global-infrastructure/regional-product-services/) for
+more information.
+
+Deploy | Region Name | Region | VPC | Bastion | DB | Fargate | Elastic Beanstalk
```
Comment (jpignata) [inline README.md]: This is very hard to parse for me, and I pretend to know what I'm doing. I think it's going to be very confusing for folks to narrow down to the right option with this mega-table.

I'd recommend at a minimum breaking this up into paths. Looking at the rows, you're working to represent at most five different paths depending upon Region:

- Basic Infrastructure (VPC)
- Basic Infrastructure (VPC) + Bastion
- Basic Infrastructure (VPC) + Bastion + App (Fargate)
- Basic Infrastructure (VPC) + Bastion + App (Fargate) + DB 
- Basic Infrastructure (VPC) + Bastion + App (Elastic Beanstalk) + DB 

Breaking these into five tables would it easier for a user to select what they need. They can first figure out which table (i.e. what infrastructure they'd like to create) and then from that table select a region. It also really calls out the composability of what you're building here.

File: README.md [hunk 2]
```diff
@@ -144,3 +144,122 @@ assigned to the "AppSecurityGroup" so they can talk to the database as
...
+:---: | ------------ | ------------- | ------------- | ------------- | -------------  | ------------- | -------------
+[🚀 ][us-east-1-vpc] | US East (N. Virginia) | us-east-1 | 👍  ||||
+[🚀 ][us-east-1-vpc-bastion] | US East (N. Virginia) | us-east-1 | 👍  | 👍 |||
+[🚀 ][us-east-1-vpc-bastion-fargate] | US East (N. Virginia) | us-east-1 | 👍  | 👍  || 👍  ||
```
Comment (jpignata) [inline README.md]: I idly wonder if you could kill the App (Fargate) without database path as you do not provide this path for Elastic Beanstalk.

File: templates/bastion.cfn.yml [hunk 3]
```diff
@@ -39,33 +42,35 @@ Mappings:
   # Note: This has not been tested with Amazon Linux 2
   AMIMap:
```
Comment (jpignata) [inline templates/bastion.cfn.yml]: The change rate of this mapping is going to be high. One thought that occurs to me is moving to an `AWS::Include` transform that holds this mapping. You could then programmatically generate the latest AMI using a script such as https://gist.github.com/jpignata/990c9bd3a7d6a4fd93b1be20bd7a0776.

This is on my list for a few templates I manage. Probably not for this PR as it's large enough already.

File: templates/elastic-beanstalk.cfn.yml [hunk 4]
```diff
@@ -331,18 +340,17 @@ Resources:
       # ENVIRONMENT VARIABLES - COMMON TO ALL STACKS
       - Namespace: aws:elasticbeanstalk:application:environment
         OptionName: AWS_REGION
-        Value: !Ref "AWS::Region"
+        Value: !Ref AWS::Region
 
-      # ENVIRONMENT VARIABLES - NODE, RAILS
+          # ENVIRONMENT VARIABLES - NODE, RAILS - Move to parameter store
```
Comment (jpignata) [inline templates/elastic-beanstalk.cfn.yml]: Is this a TODO? Accidental indent?

File: templates/elastic-beanstalk.cfn.yml [hunk 5]
```diff
@@ -3,17 +3,19 @@ Description:  AWS Elastic Beanstalk app - built by AWS Startup Kit
 
 # App stack creation prerequisites:  first create a VPC stack, then a DB stack.
 
+Description: Elastic Beanstalk
```
Comment (jpignata) [inline templates/elastic-beanstalk.cfn.yml]: These descriptions should be fleshed out using information either in the README or that you add to the README if you follow my paths advice.

File: bin/test.py [hunk 6]
```diff
@@ -0,0 +1,282 @@
+#!/usr/bin/env python
```
Comment (jpignata) [inline bin/test.py]: I could not run this script:

```console
~$ ./test.py
Testing stacks
AWS session created
Ensure key pair in region: ap-south-1
Ensure S3 app bucket in region: ap-south-1
Traceback (most recent call last):
  File "./test2.py", line 281, in <module>
    main()
  File "./test2.py", line 274, in main
    ensure_foundation(session, config)
  File "./test2.py", line 212, in ensure_foundation
    if not has_bucket(s3_client, app_bucket_name):
  File "./test2.py", line 73, in has_bucket
    raise ce
botocore.exceptions.ClientError: An error occurred (AccessDenied) when calling the GetBucketLocation operation: Access Denied
```

As such, I didn't spend too much time reading through it.

File: bin/test.py [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+TEST_APP_BUCKET_PREFIX='awslabs-startup-kit-templates-test-eb-v1-tmp-'
+TEST_APP_SOURCE_BUCKET='awslabs-startup-kit-templates-test-eb-v1'
+TEST_PYTHON_APP_KEY='eb-python-flask.zip'
+STACK_BUCKET='awslabs-startup-kit-templates-deploy-v1'
+EB_TEMPLATE_URL='https://s3.amazonaws.com/awslabs-startup-kit-templates-deploy-v1/vpc-bastion-eb-rds.cfn.yml'
+VPC_TEMPLATE_URL='https://s3.amazonaws.com/awslabs-startup-kit-templates-deploy-v1/vpc.cfn.yml'
+VPC_BASTION_TEMPLATE_URL='https://s3.amazonaws.com/awslabs-startup-kit-templates-deploy-v1/vpc-bastion.cfn.yml'
+KEY_PAIR_PREFIX='sktemplates-test-'
+
+# TODO Add support for Fargate test in supported regions
```
Comment (jpignata) [inline bin/test.py]: I wonder if it makes sense to break this down into multiple scripts that reuse the same functions, each of which have a `SUPPORTED_REGIONS` list to make it really clear what is under test and where. If you're iterating on the Fargate templates, for example, it makes not much sense to run the Elastic Beanstalk acceptance tests.

File: bin/test.py [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+# TODO Delete the RDS snapshots and any other resources
+# TODO ssh into bastion host to validate
+# TODO Make http requests to endpoints to validate
+# TODO Connect to database through bastion to validate
+# TODO Store key pair temporarily and then cleanup
```
Comment (jpignata) [inline bin/test.py]: An awful lot of TODOs on a multiple-hundred-line script further points my nose in the "decompose" direction.

File: README.md (hunk 1 above)
Comment (john) [reply README.md]: I have an updated README, really a total rewrite, coming in a separate PR--I think we need to:
- Review this PR, give it a thumbs-up (when appropriate), but hold off on merging to master
- Do same for README PR
- Merge them back-to-back, since @rnzsgh's PR will obsolete the current README, and the new README references the new material.

@koroshan has a CF template for billing alerts forthcoming. I'm going to remove the description of that from the README and add it back once Roshan's work is merged.

**Comment by john** [comment]
Can we push the test script in a subsequent PR, just for change control?

File: bin/deploy.sh [hunk 9]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
+
+#
+# Usage: ./bin/deploy.sh awslabs-startup-kit-templates-deploy-v1 startup
+#
+# The first argument is the bucket and the second is the aws cli profile
```
Comment (john) [inline bin/deploy.sh]: A comment with a short description of what this does would be useful.

**Comment by john** [comment]
This is a pre-existing thing not from this PR, but in bastion.cfn.yml, line 305, the description says it alarms on >10 in 5 minutes, but the threshold on line 311 is set to 15 (imo the threshold should be lower, maybe 5)

File: templates/db.cfn.yml [hunk 10]
```diff
@@ -1,6 +1,8 @@
 ---
 AWSTemplateFormatVersion: 2010-09-09
```
Comment (john) [inline templates/db.cfn.yml]: Nit, but consider changing the name of this file to rds-db.cfn.yml?

**Comment by rnzsgh** [comment]
@jpignata I hear you regarding the size and scope. Sometimes, you take on too much under a tight deadline and you just have to get things out the door. In the future, I will try to keep the PRs smaller and more focused. I agree about the temporary buckets and keys, and I will follow-up with another PR to rectify this situation. As @john mentioned, he has a follow-up with the README. 

@john - Let's chat offline to coordinate your readme changes. I need to get this up because there are customers who want to get going with. I agree with JP about breaking up the large tables into smaller ones, so LUD.

File: README.md (hunk 2 above)
Comment (rnzsgh) [reply README.md]: The old EB template does not support creation without DBs, but I would like to change this in the future. Often, we see startups with existing DBs who want to migrate to the Startup Kit.

File: bin/test.py (hunk 6 above)
Comment (rnzsgh) [reply bin/test.py]: Yeah, I am not currently not doing much in terms of credential validation, but I will change that in the future.

File: bin/test.py (hunk 7 above)
Comment (rnzsgh) [reply bin/test.py]: Possibly. Rich wants the foundation in and then we are going to iterate on it.

File: bin/test.py (hunk 8 above)
Comment (rnzsgh) [reply bin/test.py]: Totally. I will get these into SIM and out of the repo in the next release.

File: README.md (hunk 2 above)
Comment (jpignata) [reply README.md]: :thumbsup: Seems very reasonable to support common migration use-cases.

File: templates/bastion.cfn.yml (hunk 3 above)
Comment (rnzsgh) [reply templates/bastion.cfn.yml]: Agreed. Created an issue in the backlog.

File: templates/elastic-beanstalk.cfn.yml (hunk 5 above)
Comment (rnzsgh) [reply templates/elastic-beanstalk.cfn.yml]: They look really good the way they are in the console. The wrapper stacks look like: "VPC + Bastion + Elastic Beanstalk + Database" The nested stacks that are launched then say, "Elastic Beanstalk"

File: bin/deploy.sh (hunk 9 above)
Comment (rnzsgh) [reply bin/deploy.sh]: Added a SIM to flush out param checking and usage.

File: README.md (hunk 1 above)
Comment (rnzsgh) [reply README.md]: I added a SIM ticket and assigned to @john. I agree with breaking things down by stack, but it could turn into a series of huge tables. Does it make sense to create separate doc files, with tables for each cluster of functionality? I need to get it up the way it is now, because I have people waiting for me to send them the link.

File: bin/test.py (hunk 6 above)
Comment (rnzsgh) [reply bin/test.py]: Backlog SIM created

File: bin/test.py (hunk 7 above)
Comment (rnzsgh) [reply bin/test.py]: Backlog SIM created

File: bin/test.py (hunk 6 above)
Comment (jpignata) [reply bin/test.py]: Gotcha. FWIW, I suspect this is specifically due to my account not having access to get `GetBucketLocation` on the region-specific bucket resource.

File: templates/db.cfn.yml (hunk 10 above)
Comment (rnzsgh) [reply templates/db.cfn.yml]: I'm totally conflicted on this one. Aurora is RDS too. I am not going to address in this PR, but I will look at in the future.  SIM added for bastion change and assigned to @koroshan
